javarosa-dag.rst

```diff
@@ -0,0 +1,304 @@
+- ``QuickTriggerable`` --- A ``Triggerable`` class wrapper which provides a shallow comparison of ``Triggerble`` implementations. Once the DAG structure is built, no new ``Triggerable`` instances are created and therefore there is no need for the deep comparison. For the sake of readability, this document will always reference to ``Triggerable`` class even if ``QuickTriggerable`` is actually used in the code. 
+- ``IDag`` --- An abstract class represents an interface for the DAG management and triggerable evaluation logic.
+- ``LatestDagBase`` --- An abstract implementation of ``IDag`` class which serves as a base for all actual implementation of ``IDag``
+- ``Safe2014DagImpl`` --- The only supported ``IDag`` implementation.
+- ``FormDef`` -- A form representation along with the data model. This class communicates with ``IDag``.
+
+High level description
+-----------------------
+
+Each triggerable expression is evaluated at form load and save. Additionally, if the XPath expression references other node(s),
+it is evaluated if any of the referenced node is updated. Moreover, it may happen that a referenced node also has a triggerable
+expression which references other nodes etc. 
+So triggerables may depend on each other. 
+
+Lets check an example:
+Assume there are nodes ``nodeA``, ``nodeB`` and ``nodeC`` and:
+
+* ``nodeA`` has a ``calculate`` expression ``tA`` which references the ``nodeB`` value
+* ``nodeB`` has a ``calculate`` expression ``tB`` which references the ``nodeC`` value
+* ``nodeC`` has no triggerable expressions
+
+So the dependency is ``tC`` > ``tB`` > ``tA``.
+When a form is loaded or saved, those triggerables should be evaluated in the following order: ``tC``, ``tB``, ``tA``.
+When a value of ``nodeC`` is changed then ``tB`` and ``tA`` must be evaluated. In this order.
+
+Lets change the above so the ``nodeC`` node has a ``relevant`` expression which returns ``true`` if ``nodeA`` has certain value.
+This gives the following dependency ``tA`` > ``tC`` > ``tB`` > ``tA``. Since ``tA`` depends on ``tB`` and ``tB`` depends on ``tC``
+it results in an infinity loop. A paradox which cannot be handled.
+
+Direct Acyclic Graph ensures that triggerables are in the topological order. If there are cycles then DAG cannot be constructed
+and an exception is thrown.
+
+``Triggerable``
+----------------
+
+This class stores the following information:
+
+* The expression to be evaluated.
+* A list of target nodes which should be updated with the expression evaluation result.
+* The context reference.
+* A list of ``Triggerable`` instances - named ``immediateCascades`` - that depend on ``this`` evaluation result and should be evaluated after ``this`` is evaluated.
+
+Note: terms 'triggerable', 'calculation' or 'condition' mean the same - an instance of ``Triggerable`` class - and are used interchangeably in this document.
```

Comment:
em dashes are great fun! Tough to make on a typewriter but not today.

mean the same—an instance … class—and are